### PR TITLE
SpaceX Group Project

### PR DESCRIPTION
**Pull request by Elerqsousy**
## Build Space Travelers' Hub Group Project using React and Redux Toolkit

The work was done collaboratively by a team of two players:
1. [Riez](https://github.com/Elerqsousy): was responsible for Rockets view
2. [Ekene](https://github.com/ekenecf): was responsible for the Missions view

<hr>

In this Project:
1. Setup project dependencies and Linters:
- Used Create-React-App to initiate the project
- Installed [React Redux](https://react-redux.js.org/), [Redux Logger](https://www.npmjs.com/package/redux-logger) and [React Router](https://reactrouter.com/web/guides/quick-start).
- configured Linters for React-Redux project
- Deployed project Dev branch on Netlify
- Updated Readme Authors and Intro sections, and Live Demo Link.
2. Build Project Header
- For this part, I created the project header with navigation inside.
3. Create routes
- I create a route and a view component. 
- Use <NavLink /> for the page navigation links and style active class to indicate which section/page the user is currently on.
4. Build Rocket List Structure
- Used useSelector(), Redux Hook, to select the state slices and render lists of rockets in corresponding routes.
- Used  [React Bootstrap](https://react-bootstrap.github.io/) library to temporarily style the view.
5. Created the missions display
- Rendered a list of all joined missions
- Dispatched actions upon clicking on the corresponding buttons.
- Created a route and a view component.
- Used useSelector(), Redux Hook, to select the state slices and render lists of missions in corresponding routes.
- Implemented a function such that when a user clicks the "Join Mission" button, action needs to be dispatched to update the store.
6. Build Rocket page structure and data Fetching
- Created a route and a view component for the Rockets page. 
- Used `<NavLink />` for the page navigation link 
- Used Axios, Redux ToolKit, and local storage for managing data
- Upon application first load: (Fetch data from the [Rockets endpoint](https://api.spacexdata.com/v3/rockets), data from API gets handled and only id, name, type, flickr_images get passed to the redux store + a property set to false for reservation tracking, local storage gets populated with updated data)
- API end point only gets populated once in the whole app lifecycle.
7. Build Rocket List Structure
- Used useSelector(), Redux Hook, to select the state slices and render lists of rockets in corresponding routes.
- Used  [React Bootstrap](https://react-bootstrap.github.io/) library to temporarily style the view.
8. Subscribe Rocket reservation to Redux store
 Added reservation reducer and action creator to the rockets slice
- When dispatching reservation action it toggles reserved property for specified list item
- according to the reserved property of list Item: (the badge conditionally renders, the button text and functionality change from reserve to cancel the reservation)
9. Filter and Display Rockets 
- Filtered the reserved rocket list
- Displayed the filtered list on My profile page
10. Add Extra Features to My Profile Page (Rockets)
- Added a placeholder message when the "My Rockets" lists are empty.
- Added "Leave Mission" buttons to the list of reserved rockets on my profile. Clicking them dispatch unreserve rocket action.
- Added the "Read more" button for each rocket. Upon click, it should open a corresponding Wikipedia page in a new tab using URL from API.
11. Styling Pages
- Used CSS modules
- Added Rocket page and Profile Pages styling
- Added Nav styling and quick fixes
- Added missions page styling

<hr>

Project deployed using [Netlify](https://gorgeous-melba-746edd.netlify.app/).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,12 +3,26 @@
...
+    "prop-types": "^15.8.1",
     "react": "^18.1.0",
+    "react-bootstrap": "^2.4.0",
     "react-dom": "^18.1.0",
+    "react-redux": "^8.0.2",
+    "react-router-dom": "^6.3.0",
     "react-scripts": "5.0.1",
+    "react-toastify": "^9.0.1",
+    "redux": "^4.2.0",
+    "redux-logger": "^3.0.6",
```
Comment (Cathella) [inline package.json:21]: ✅ [Redux Logger](https://www.npmjs.com/package/redux-logger) should exist on the package.json file 👍🏻

File: package.json [hunk 2]
```diff
@@ -3,12 +3,26 @@
...
     "@testing-library/react": "^13.2.0",
     "@testing-library/user-event": "^13.5.0",
+    "axios": "^0.27.2",
+    "bootstrap": "^5.1.3",
+    "prop-types": "^15.8.1",
     "react": "^18.1.0",
+    "react-bootstrap": "^2.4.0",
     "react-dom": "^18.1.0",
+    "react-redux": "^8.0.2",
+    "react-router-dom": "^6.3.0",
```
Comment (Cathella) [inline package.json:17]: ✅ [React Router](https://reactrouter.com/web/guides/quick-start) should exist on the package.json file 👍🏻

File: package.json [hunk 3]
```diff
@@ -3,12 +3,26 @@
...
     "@testing-library/jest-dom": "^5.16.4",
     "@testing-library/react": "^13.2.0",
     "@testing-library/user-event": "^13.5.0",
+    "axios": "^0.27.2",
+    "bootstrap": "^5.1.3",
+    "prop-types": "^15.8.1",
     "react": "^18.1.0",
+    "react-bootstrap": "^2.4.0",
     "react-dom": "^18.1.0",
+    "react-redux": "^8.0.2",
```
Comment (Cathella) [inline package.json:16]: ✅ [React Redux](https://react-redux.js.org/) should exist on the package.json file 👍🏻

File: src/App.js [hunk 4]
```diff
@@ -1,24 +1,22 @@
...
-          Learn React
-        </a>
-      </header>
-    </div>
+    <Router>
+      <Nav />
+      <Routes>
+        <Route path="/" element={<Rockets />} />
+        <Route path="/profile" element={<Profile />} />
+        <Route path="/mission" element={<Missions />} />
```
Comment (Cathella) [inline src/App.js:17]: ✅ Proper routes and view components should be created (Rockets, Missions, My Profile, 👍🏻

File: src/navigation/Nav.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      <div className="left-nav">
+        <img src={logo} alt="logo" />
+        <h1>Space Travelers&apos; Hub</h1>
+      </div>
+      <nav className="right-nav">
+        <ul className="nav-links">
+          {routLinks.map((rout) => (
+            <li key={uuid()} className="nav-link">
+              <NavLink to={rout.path}>{rout.text}</NavLink>
+            </li>
```
Comment (Cathella) [inline src/navigation/Nav.js:32]: ✅ `<NavLink />` should be used for navigation links 👍🏻

File: src/navigation/Nav.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      text: 'Rockets',
+    },
+    {
+      path: '/mission',
+      text: 'Missions',
+    },
+    {
+      path: '/profile',
+      text: 'My Profile',
+    },
```
Comment (Cathella) [inline src/navigation/Nav.js:19]: - [ ] [REQUIRED] Only the active page should be indicated with an underline. Kindly implement this.

File: src/pages/Profile.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import React, { useEffect, useState } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import updateOnLoad from '../components/onLoad';
+import ReservedRocketList from '../components/ReservedRocketList';
+import api from '../reduxFiles/api';
+import local from '../reduxFiles/local';
+import styles from '../styles.module.css';
+
+const Profile = () => {
+  const rockets = useSelector((state) => state.rockets);
```
Comment (Cathella) [inline src/pages/Profile.js:10]: ✅ `useSelector()` should be used to get a state slice corresponding to routes 👍🏻

**Review by Cathella — CHANGES_REQUESTED** [review]
## 🟡 Status: Changes Required 🛑

### Hi @Elerqsousy , @ekenecf 👋🏻, 

Good job so far! The project looks great 😍, especially the missions part - is visually attractive. We need more styling work done on the rockets part so that it matches the required design in the requirements.
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## To Highlight 👏🏻

✅ No linter errors.
✅ You used correct GitFlow.
✅ Fetches data and update Redux store.

## Required Changes ♻️
_Check the comments under the review._

- [ ] [REQUIRED] Kindly style the missions table and call-to=actions to match the design below

image


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by DeliceLydia — APPROVED** [review]
## Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
# Congratulations! 🎉 



### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**